### DISCUSSION TITLE
Immich License Changes

### DISCUSSION THREAD

**Discussion by alextran1502** [immich-app/immich#7023] [Announcements]
# Immich License Changes

> [!IMPORTANT]
> **Immich is switching from MIT to the AGPLv3 License**.

We are changing the license to protect Immich, its community, and the effort the contributors have put in to develop the application over the past two years.

### Why AGPLv3?

In simple terms, we want to keep anyone from using the Immich source code without making their changes public or contributing back to the project. That includes ourselves, so you can rest assured that Immich will always stay free and open-source.

<img width="600px" alt="image" src="https://github.com/immich-app/immich/assets/27055614/c47a0d65-574d-4989-b1f3-1b3a0261e7ec">
    
### What does this mean for Immich? Is it going down the dark path?
    
No, Immich is always guaranteed to be free and open-source and will never have any paywalled features. All features will be available for all users.

### What will happen to services like PikaPods or Cloudron
    
Services that offer Immich on their platform can continue to do that. If any of them modify Immich's source code, they will need to open their modifications of the source code under the same AGPL license and include a prominent notice to their users stating where this source code can be found.

- - - -

Immich Team

**Comment by rathmerdominik** [comment]
A very smart choice, that i can agree with❤️
In no way should greed be involved in Projects like these.

**Comment by PrivateGER** [comment]
Very good decision. Don't let your work be stolen by random companies.

**Comment by ndragon798** [comment]
Great to see the license changing for the better. :)

**Comment by Defective4467** [comment]
Very good choice! Living it

**Comment by A-wels** [comment]
I really like this change! 

Out of interest: If one of the named companies decides to stay on the last MIT Version, would this change still apply to them?

**Reply by gzz2000** [reply]
I guess unfortunately no, but they will not able to use new features/bugfixes thereafter without complying to AGPL.

**Comment by ssddanbrown** [comment]
With this change will you be requiring contributors to sign a CLA? Or anything similar that retains rights of re-licensing to yourself?

**Reply by bo0tzz** [reply]
This change was prompted in part because people expressed concerns that we might pull a bait-and-switch and suddenly start charging for (parts of) Immich. Moving to AGPL removes that possibility, but requiring a CLA would bring it back, defeating the point. As such we will not be requiring a CLA.

**Reply by ssddanbrown** [reply]
Awesome! Yeah, that's what I was double checking. Much respect to the commitment of AGPL without CLA!

**Comment by nebulade** [comment]
Hi, Cloudron co-founder here. We too really like the license change. Anyways we don't maintain cloudron specific changes for immich or any app for that matter. We always upstream our changes if any. Our business model is only around making deployment of the app simpler.

**Comment by Venthe** [comment]
Very sad day. Instead of code being free, you impose politics with it. I for one would not be interested in committing to xGPL licensed code.

**Reply by PrivateGER** [reply]
The AGPL is not politics, it's the strongest copyleft license available and supported by the GNU project, one of the THE biggest supporters of open source software.

**Reply by ssddanbrown** [reply]
> Instead of code being free

The AGPL, compared to other licenses, does a particularly very good job at ensure the code is, and remains, free to it's users. 
Especially so where there's not CLA/handover-of-rights like authors have committed to here, since the copyright and right-to-relicense is shared among its contributors.

**Reply by Defective4467** [reply]
That's a silly perspective, it's a change to mean you don't get to monetize someone else's work. If your in opposition are you speaking to your intent to do just that, or is there another reason?

**Reply by istankovic** [reply]
> Very sad day. Instead of code being free, you impose politics with it. 

Many, many things are unavoidably political.
Especially free software licenses. A license specifies which behaviour is acceptable and which isn't.

The fact that most people these days choose to just ignore "the license question" and pick whatever is trending (e.g. MIT, Apache, BSD...) is rather disappointing. Would you say that people choosing MIT avoid politics? IMO, they're not avoiding politics at all. (Whether they are actually aware of that is a different question.) They're merely choosing a specific policy, one that differs from GPL and other licences.

Now, I'm sure you would agree that whoever writes the code also has the right to choose the license, and thereby
the ways in which that code can be used. Saying that choosing a particular license is politics, but choosing another one isn't,
sounds a bit like saying that water is a liquid, but wine isn't. You may prefer one flavor over the other, and you decide what you want to drink. The choice of what comes to the table is, however, with others.

> I for one would not be interested in committing to xGPL licensed code.

What does "committing" here mean? Could you be more specific, please?
Did you mean to say, "I am not interested in contributing to xGPL code"?
Or did you mean to say, "I am not interested in using xGPL code"?

**Reply by ssddanbrown** [reply]
@Defective4467 even monetizing immich would still be allowed under the AGPL. That's a fundamental right of free and open source software. It's just harder to do in a way that's modifying & selling the software itself since any customers would have the right to freely redistribute the source to others (and potentially copy changes back into the main immich code or other forks where desired).

**Reply by Venthe** [reply]
Sorry, I was quite away from the GH.

@PrivateGER 

> The AGPL is not politics, it's the strongest copyleft license (...)

Copyleft is inherently political, as it limits the usage of the code from the licensing standpoint.

@ssddanbrown

> The AGPL, compared to other licenses, does a particularly very good job at ensure the code is, and remains, free to it's users.
Especially so where there's not CLA/handover-of-rights like authors have committed to here, since the copyright and right-to-relicense is shared among its contributors.

The only thing it ensures with the context of Immich is that the original authors cannot relicense the code again without approval of the commiters/removal of their contribution. Permissive licenses like MIT allow literally the same thing for Immich users - as the code is free, open, forkable - with the additional benefit of not clamping down on the downstream usage in the name of the politics.

@Defective4467 

> don't get to monetize someone else's work. If your in opposition are you speaking to your intent to do just that, or is there another reason?

Another. I am fundamentally against political and limiting copyleft, cancerous licenses. Open source means open, not Open* (but others are limited in what they can do).

If I produce an open source code (And I've done my share), it is open to anyone to do what they please. I don't attach politics and requirements to it.

@istankovic

> Many, many things are unavoidably political. Especially free software licenses. A license specifies which behaviour is acceptable and which isn't. (...) Now, I'm sure you would agree that whoever writes the code also has the right to choose the license, and thereby
the ways in which that code can be used.

True. That's why I've put my voice out; regardless if people agree with it or not.

> Would you say that people choosing MIT avoid politics?

Depends. I'd say that people choosing specifically copyleft licensing are political.

> What does "committing" here mean? Could you be more specific, please?
> Did you mean to say, "I am not interested in contributing to xGPL code"?
> Or did you mean to say, "I am not interested in using xGPL code"?

Sure.  "I am not interested in contributing to xGPL code". I don't care about the code licensing as a user, I do care as a developer. Either code is free or it isn't, and copyleft like xGPL limit how the code can be distributed; limit further re-licensing. If I'm doing open source, I will not take part in something that limits someone else's freedom - in any shape or form.

---

The issue is, I am _really_ pro-open source, and pro-code that is _actually_ free. If that means that someone else's wishes to profit from it, I don't care - it only means that I could do the same yet decided I won't. If they do a better job with a paid product, then that's a net benefit for everyone. If they monetize, yet the original code is still leading the way, I've lost nothing. If someone re-licensed the code, original code is still there. It is, again - *free* for everyone for anything. No politics attached, no ideologies and no requirements.

**Reply by bo0tzz** [reply]
> The only thing it ensures with the context of Immich is that the original authors cannot relicense the code again without approval of the commiters/removal of their contribution.

This was a deliberate part of the decision. With MIT, it would be possible to do that, which means there would be a possibility to unilaterally change the license in the future to something that negatively impacts the community. That is now no longer an option. This change not only guarantees that the code _is_ free, but also that it _stays_ free.

**Reply by Venthe** [reply]
The issue is, GPL forces that decision on any future contributor.

Let us consider two scenarios:

1) mit->closed

People will either fork it and do a better job; or paid option will provide better value. Either way, consumers win. Code stays free and open; both for future re-licensing and open source development. Nothing is imposed.

2) mit->copyleft

Results are no longer open to every party. Developing a paid product based on it is risky and unviable. In this case, the only real development can happen and can die with the gpl. Code will not be forked due to the illusion of openness. Ultimately, this path closes off part of the possible development.

---

From my perspective, mit->GPL here is a loss-loss for everybody.

E: but that's completely besides the point. Immich, which i like decided to make a decision which in my opinion is detrimental. They intend to do it anyway, so any further discussion is moot. Still, thanks for swapping thoughts

**Reply by 1e100** [reply]
I can see why a "business consultant" would be displeased with AGPL - businesses prefer to stay away from copyleft licenses. But I don't see how that displeasure is relevant to most of us here. You're free to fork and maintain the version before the license change, I think, if you feel so strongly about the issue.

**Reply by Venthe** [reply]
My profession has nothing to do with this; as I've said, it's about freedom which copyleft is ultimately against - and specifically this project, which change to the licensing in my opinion will be detrimental to it 

Anything else does not belong to this discussion, so either we stick to this project and this change in particular; or I welcome you to a discussion elsewhere.

**Reply by istankovic** [reply]
I think the issue that @Venthe is having with this change is simply due to different values, which are obviously not shared by the rest of the community welcoming this change. When the values are not aligned, there is not much hope of progress.

Also, as @1e100 said, anyone is free to fork the code before the change and do whatever they want. Perhaps a new community would gather around the (supposedly non-political) license? Who knows, if enough people are interested... :shrug:

**Reply by Venthe** [reply]
Precisely. That's why I'm not going to push the issue any further. The time will tell if this was a good decision regardless of the sentiment.

**Comment by 1e100** [comment]
This makes me feel pretty good about my digital future - now I can be certain Immich will remain free as in freedom indefinitely, rather than subverted by commercial interests at some point (and what a ripe target it was for that). This is specifically true because of the no-CLA stance. This means the team cannot unilaterally re-license the code at a later date, effectively ensuring freedom and denying future interference.

**Comment by dirkriehle** [comment]
Politics or not, here is my hopefully sober take on why folks relicense permissive->copyleft or copyleft->permissive. https://dirkriehle.com/2024/02/12/how-open-source-licenses-increase-or-curtail-reach-of-the-software/

**Comment by zvang2** [comment]
알려줘서 고마워. 업데이트에 감사하고 지원합니다. 행복한 코딩 커뮤니티 우리 모두는 항상 커뮤니티를 더 좋게 만들 수 있는 해결책을 위해 노력해 왔습니다.

Thank you for letting me know Thank you for the update and support it. Happy Coding Community We've all always been working on solutions that can make our community better.

💕

**Comment by raffomania** [comment]
Good choice! There's been uncertainty about uploading AGPL-licensed apps into Apple's app store before, do you plan on dual-licensing the Immich iOS app for that purpose?

**Reply by ajayyy** [reply]
You can do what nextcloud does https://github.com/nextcloud/ios/blob/master/COPYING.iOS

No dual licensing needed

**Reply by alextran1502** [reply]
No CLA is required for Immich

**Comment by jakob42** [comment]
Just out of curiosity: Does one need support from all commiters so far to change the license? I welcome the change, good step 👍

**Reply by ssddanbrown** [reply]
No, not in this particular change since the project has been under the MIT license which allows sub-licensing, so it's fine as long as the rights provided of the old license are still respected (Attribution of copyright and maybe reference/include the MIT license of the original work for reference) and the new license is compatible with those rights (which it is in this case).

If you wanted to change from AGPL however, then that license does not allow sub-licensing (although there's technically some compatibility with the GPLv3 but that's slightly different). You'd need permission from all those that have contributed while the project was under that AGPL license (some projects get this at time of contribution, commonly via a "CLA").

Just to confirm, I am not a legal expert nor affiliated with the Immich project.

**Reply by dirkriehle** [reply]
You can't retroactively change a license so the old code will remain under the original MIT license.

Only the copyright holder can change the license so if there is only @alextran1502 they can put the new license on the old code. This does not invalidate the prior license for this old code.

For any new code, only the new license applies.

For files that mix old and new code, the new license wins for the full file. You could still copy over to other places that part of the file that remained untouched and hence available under the old license.

See also reuse.software to ensure licensing is clear on a file level.

**Comment by avinal** [comment]
Great move. Ensuring the future of the project and its dedicated users is always welcome.

**Comment by raihanhossainbd** [comment]
Great decision ever👌🔥. It is more secured than ago✊.Thanks a lot Immich authority🌺 Love from🇧🇩
